src/core/execute.c

```diff
@@ -2244,12 +2252,54 @@ static bool exec_directory_is_private(const ExecContext *context, ExecDirectoryT
         return true;
 }
 
+/* StateDirectorySymlinks and friends should reference only existing StateDirectory sources */
+static int prune_non_existing_exec_directories_symlinks(
+                const ExecContext *context,
+                const ExecParameters *params,
+                ExecDirectoryType type) {
+
+        int r;
+
+        assert(context);
+        assert(params);
+        assert(type >= 0 && type < _EXEC_DIRECTORY_TYPE_MAX);
+
+        if (!params->prefix[type])
+                return 0;
+
+        char **src, **dst;
+        STRV_FOREACH_PAIR(src, dst, context->directories[type].symlinks) {
+                /* The source has to be already specified in the corresponding directory type list,
+                 * otherwise mark it for pruning. */
+                if (strv_contains(context->directories[type].paths, *src))
+                        continue;
+
+                log_warning("Source directory in %s=%s:%s tuple was not part of %s=, ignorning.",
+                                exec_directory_type_symlink_to_string(type),
+                                *src,
+                                *dst,
+                                exec_directory_type_to_string(type));
+
+                r = free_and_strdup(src, "");
+                if (r < 0)
+                        return -ENOMEM;
+                r = free_and_strdup(dst, "");
+                if (r < 0)
+                        return -ENOMEM;
+        }
```

Comment:
I'd just write this like this:

```c
for (size_t i = 0, j = 0; context->directories[type].symlinks[i]; i += 2) {
        assert(context->directories[type].symlinks[i+1]);
        if (…) {
                 context->directories[type].symlinks[j++] = context->directories[type].symlinks[i];
                 context->directories[type].symlinks[j++] = context->directories[type].symlinks[i+1];
                 continue;
        }
        …

        free(context->directories[type].symlinks[i]);
        free(context->directories[type].symlinks[i+1]);
}

context->directories[type].symlinks[j] = NULL;
```

i.e. iterate with a source and a target index, and then copy source to target if good, and free source otherwise. Only increase target index in the former case.

this means you won#t have to hack in memory allocations.